pkg/restore/merge.go

```diff
@@ -0,0 +1,144 @@
+	DefaultMergeRegionSizeBytes uint64 = 96 * utils.MB
+
+	// DefaultMergeRegionKeyCount is the default region key count, 960000.
+	DefaultMergeRegionKeyCount uint64 = 960000
+
+	writeCFName   = "write"
+	defaultCFName = "default"
+)
+
+// MergeRangesStat holds statistics for the MergeRanges.
+type MergeRangesStat struct {
+	TotalFiles           int
+	TotalWriteCFFile     int
+	TotalDefaultCFFile   int
+	TotalRegions         int
+	RegionKeysAvg        int
+	RegionBytesAvg       int
+	MergedRegions        int
+	MergedRegionKeysAvg  int
+	MergedRegionBytesAvg int
+}
+
+// MergeFileRanges returns ranges of the files are merged based on
+// splitSizeBytes and splitKeyCount.
+//
+// By merging small ranges, it speeds up restoring a backup that contains many
+// small ranges (regions) as it reduces split region and scatter region.
+func MergeFileRanges(
+	files []*kvproto.File, splitSizeBytes, splitKeyCount uint64,
+) ([]rtree.Range, *MergeRangesStat, error) {
+	if len(files) == 0 {
+		return []rtree.Range{}, &MergeRangesStat{}, nil
+	}
+	totalBytes := uint64(0)
+	totalKvs := uint64(0)
+	totalFiles := len(files)
+	writeCFFile := 0
+	defaultCFFile := 0
+
+	filesMap := make(map[string][]*kvproto.File)
+	for _, file := range files {
+		filesMap[string(file.StartKey)] = append(filesMap[string(file.StartKey)], file)
+
+		// We skips all default cf files because we don't range overlap.
+		if file.Cf == writeCFName || strings.Contains(file.GetName(), writeCFName) {
+			writeCFFile++
+		} else if file.Cf == defaultCFName || strings.Contains(file.GetName(), defaultCFName) {
+			defaultCFFile++
+		}
+		totalBytes += file.TotalBytes
+		totalKvs += file.TotalKvs
+	}
+	if writeCFFile == 0 && defaultCFFile == 0 {
+		return []rtree.Range{}, nil, errors.Annotatef(berrors.ErrRestoreInvalidBackup,
+			"unknown backup data from neither Wrtie CF nor Default CF")
+	}
+
+	// RawKV does not have data in write CF.
+	totalRegions := writeCFFile
+	if defaultCFFile > writeCFFile {
+		totalRegions = defaultCFFile
+	}
+
+	// Check if files are overlapped
+	rangeTree := rtree.NewRangeTree()
+	for key := range filesMap {
+		files := filesMap[key]
+		if out := rangeTree.InsertRange(rtree.Range{
+			StartKey: files[0].GetStartKey(),
+			EndKey:   files[0].GetEndKey(),
+			Files:    files,
+		}); out != nil {
+			return nil, nil, errors.Annotatef(berrors.ErrRestoreInvalidRange,
+				"duplicate range %s files %+v", out, files)
+		}
+	}
+
+	needMerge := func(left, right *rtree.Range) bool {
+		leftBytes, leftKeys := left.BytesAndKeys()
+		rightBytes, rightKeys := right.BytesAndKeys()
+		if rightBytes == 0 {
+			return true
+		}
+		if leftBytes+rightBytes > splitSizeBytes {
+			return false
+		}
+		if leftKeys+rightKeys > splitKeyCount {
+			return false
+		}
+		// Do not merge ranges in different tables.
+		if tablecodec.DecodeTableID(kv.Key(left.StartKey)) != tablecodec.DecodeTableID(kv.Key(right.StartKey)) {
```

Comment:
Can we merge index range and row key range with same table id here?

Reply:
It's ok as long as they belong to the same table.

Reply:
If we merge record key and index key into one file. we need to find the correct rewrite rule. https://github.com/pingcap/br/blob/c344d1edf20cf17d353b9d8b789919fd1172039f/pkg/restore/import.go#L212
but default rewrite rule starts with record prefix 
https://github.com/pingcap/br/blob/c344d1edf20cf17d353b9d8b789919fd1172039f/pkg/restore/util.go#L73
and with this rule sst_importer cannot rewrite index key correctly.

Reply:
You are right, we need to separate ranges by index ID too.